File: ruby/easy/reverse_bits.rb
# @param {Integer} n, a positive integer
# @return {Integer}
def reverse_bits(n)
    return reverse(0, n, { mul: 1 })
end

def reverse(i, n, obj)
    return 0 if i >= 32
    res = reverse(i+1, n >> 1, obj) + (((n & 1) != 0) ? obj[:mul] : 0)
    obj[:mul] *= 2
    return res
end

File: javascript/easy/reverseBits.js
/**
 * @param {number} n - a positive integer
 * @return {number} - a positive integer
 */
var reverseBits = function(n) {
    let stack = [], i = 0, res = 0, mul = 1;
    while(i < 32) {
        stack.push(((n & 1) != 0) ? 1 : 0);
        n >>= 1;
        i++;
    } 
    
    while(stack.length > 0) {
        res = res + (stack.pop() * mul);
        mul *= 2;
    }    
    
    return res;
};